File: eHour-persistence/src/main/java/net/rrm/ehour/persistence/project/dao/ProjectDaoHibernateImpl.java
/*
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

package net.rrm.ehour.persistence.project.dao;

import com.google.common.base.Optional;
import net.rrm.ehour.domain.Customer;
import net.rrm.ehour.domain.Project;
import net.rrm.ehour.domain.User;
import net.rrm.ehour.persistence.dao.AbstractGenericDaoHibernateImpl;
import org.hibernate.Criteria;
import org.hibernate.criterion.Restrictions;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository("projectDao")
public class ProjectDaoHibernateImpl extends AbstractGenericDaoHibernateImpl<Project, Integer> implements ProjectDao {
    protected static final Optional<String> CACHEREGION = Optional.of("query.Project");

    public ProjectDaoHibernateImpl() {
        super(Project.class);
    }

    @Override
    @SuppressWarnings("unchecked")
    public List<Project> findAllActive() {
        return findByNamedQuery("Project.findAllActive", CACHEREGION);
    }

    @Override
    @SuppressWarnings("unchecked")
    public List<Project> findDefaultProjects() {
        return findByNamedQuery("Project.findAllActiveDefault", CACHEREGION);
    }

    @Override
    public List<Project> findProjectForCustomers(List<Customer> customers, boolean onlyActive) {
        String hqlName = onlyActive ? "Project.findActiveProjectsForCustomers" : "Project.findAllProjectsForCustomers";

        return findByNamedQueryAndNamedParam(hqlName, "customers", customers.toArray(), CACHEREGION);
    }

    @Override
    public List<Project> findActiveProjectsWhereUserIsPM(User user) {
        return findByNamedQueryAndNamedParam("Project.findActiveProjectsWhereUserIsPM", "user", user, CACHEREGION);
    }

    @SuppressWarnings("unchecked")
    @Override
    public List<Project> findAllProjectsWithPmSet() {
        Criteria criteria = getSession().createCriteria(Project.class);
        criteria.add(Restrictions.isNotNull("projectManager"));

        return (List<Project>) criteria.list();
    }
}

File: eHour-persistence/src/main/scala/net/rrm/ehour/persistence/project/dao/ProjectDaoHibernateImpl.scala
package net.rrm.ehour.persistence.project.dao

import java.util

import net.rrm.ehour.domain.{Customer, Project, User}
import net.rrm.ehour.persistence.dao.AbstractGenericDaoHibernateImpl
import net.rrm.ehour.persistence.retry.ExponentialBackoffRetryPolicy
import org.hibernate.FetchMode
import org.hibernate.criterion.Restrictions
import org.springframework.stereotype.Repository
import org.springframework.transaction.annotation.Transactional

@Repository("projectDao")
class ProjectDaoHibernateImpl extends AbstractGenericDaoHibernateImpl[Integer, Project](classOf[Project]) with ProjectDao {
  private final val CacheRegion = Some("query.Project")

  @Transactional(readOnly = true)
  override def findAllActive(): util.List[Project] = findByNamedQuery("Project.findAllActive", CacheRegion)

  @Transactional(readOnly = true)
  override def findDefaultProjects(): util.List[Project] = findByNamedQuery("Project.findAllActiveDefault", CacheRegion)

  @Transactional(readOnly = true)
  override def findProjectForCustomers(customers: util.List[Customer], onlyActive: Boolean): util.List[Project] = {
    val hqlName = if (onlyActive) "Project.findActiveProjectsForCustomers" else "Project.findAllProjectsForCustomers"
    findByNamedQuery(hqlName, "customers", customers, CacheRegion)
  }

  @Transactional(readOnly = true)
  override def findActiveProjectsWhereUserIsPM(user: User): util.List[Project] = findByNamedQuery("Project.findActiveProjectsWhereUserIsPM", "user", user, CacheRegion)

  @Transactional(readOnly = true)
  override def findAllProjectsWithPmSet(): util.List[Project] = {
    val criteria = getSession.createCriteria(classOf[Project])
    criteria.add(Restrictions.isNotNull("projectManager"))
    criteria.setFetchMode("projectManager.userRoles", FetchMode.JOIN)
    ExponentialBackoffRetryPolicy retry criteria.list.asInstanceOf[util.List[Project]]
  }
}


